cmd/tracegen/main.go

```diff
@@ -19,106 +19,56 @@ package main
 
 import (
 	"context"
-	"errors"
 	"flag"
 	"fmt"
 	"log"
-	"math"
 	"math/rand"
 	"os"
 	"os/signal"
 
 	"go.elastic.co/apm/v2"
 	"go.uber.org/zap"
-	"go.uber.org/zap/zapcore"
-	"go.uber.org/zap/zapgrpc"
-	"google.golang.org/grpc/grpclog"
 
 	"github.com/elastic/apm-tools/pkg/tracegen"
 )
 
 func main() {
-	var cfg tracegen.Config
-
-	flag.StringVar(&cfg.APMServerURL, "server", "", "set APM Server URL (env value ELASTIC_APM_SERVER_URL)")
-	flag.StringVar(&cfg.APIKey, "api-key", "", "set APM API key for auth (env value ELASTIC_APM_API_KEY)")
-	flag.Float64Var(&cfg.SampleRate, "sample-rate", 1.0, "set the sample rate. allowed value: min: 0.0001, max: 1.000")
-	flag.StringVar(&cfg.OTLPProtocol, "otlp-protocol", "grpc", "set OTLP transport protocol to one of: grpc (default), http/protobuf")
-	flag.BoolVar(&cfg.Insecure, "insecure", false, "skip the server's TLS certificate verification")
-	logLevel := zap.LevelFlag(
-		"loglevel", zapcore.InfoLevel,
-		"set log level to one of: DEBUG, INFO (default), WARN, ERROR, DPANIC, PANIC, FATAL",
-	)
-
+	serverURL := flag.String("server", "", "set APM Server URL (env value ELASTIC_APM_SERVER_URL)")
+	apiKey := flag.String("api-key", "", "set APM API key for auth (env value ELASTIC_APM_API_KEY)")
+	sr := flag.Float64("sample-rate", 1.0, "set the sample rate. allowed value: min: 0.0001, max: 1.000")
+	insecure := flag.Bool("insecure", false, "sets agents to skip the server's TLS certificate verification")
+	protocol := flag.String("otlp-protocol", "grpc", "set OTLP transport protocol to one of: grpc (default), http/protobuf")
 	flag.Parse()
-	// get or set env values for GO agent and otel agent
-	err := configureEnv(&cfg)
-	if cfg.SampleRate < 0.0001 || cfg.SampleRate > 1.0 {
-		log.Fatalf("invalid sample rate %f provided. allowed value: 0.0001 <= sample-rate <= 1.0", cfg.SampleRate)
-	}
 
-	cfg.SampleRate = math.Round(cfg.SampleRate*10000) / 10000
+	apmTracer, err := apm.NewTracer(newUniqueServiceName("service", "intake"), "0.0.1")
 	if err != nil {
-		log.Fatalf("Invalid configuration: %v", err)
+		log.Fatal("failed to instantiate apm tracer")
 	}
 
-	// set logger
-	zapcfg := zap.NewProductionConfig()
-	zapcfg.EncoderConfig.EncodeTime = zapcore.RFC3339TimeEncoder
-	zapcfg.EncoderConfig.EncodeLevel = zapcore.CapitalColorLevelEncoder
-	zapcfg.Encoding = "console"
-	zapcfg.Level = zap.NewAtomicLevelAt(*logLevel)
-	logger, err := zapcfg.Build()
+	cfg := tracegen.NewConfig(
+		tracegen.WithAPMServerURL(*serverURL),
+		tracegen.WithAPIKey(*apiKey),
+		tracegen.WithSampleRate(*sr),
+		tracegen.WithInsecureConn(*insecure),
+		tracegen.WithElasticAPMTracer(apmTracer),
+		tracegen.WithOTLPProtocol(*protocol),
+		tracegen.WithOTLPServiceName(newUniqueServiceName("service", "otlp")),
+	)
+
+	err = cfg.Validate()
```

Comment:
After the changes are made, it's not necessary to explicitly validate the config, since the functions that use it will automatically do it.